src/libmongoc/src/mongoc/mongoc-client-side-encryption.c

```diff
@@ -2672,8 +2675,222 @@ _mongoc_cse_is_enabled (mongoc_client_t *client)
+   bson_destroy (&in_encryptedFields);
+   // Destroy the local options, which may or may not have been used. If unused,
+   // the new options are now owned by the caller and this is a no-op.
+   bson_destroy (&local_new_options);
+   // The resulting collection, or NULL on error:
+   return ret;
+}
+
 #endif /* MONGOC_ENABLE_CLIENT_SIDE_ENCRYPTION */
 
+/// Generate one encryptedField element.
+static void
+_init_1_encryptedField (bson_t *out_field,
+                        const bson_t *in_field,
+                        auto_datakey_factory fac,
+                        void *fac_userdata,
+                        bson_error_t *error)
+{
+   bsonVisitEach (
+      *in_field,
+      // If it is not a "keyId":null element, just copy it to the output.
+      if (not(keyWithType ("keyId", null)),
+          then (appendTo (*out_field), continue)),
+      // Otherwise:
+      do({
+         // Set up factory context
+         bson_value_t new_key = {0};
+         struct auto_datakey_context ctx = {
+            .out_keyid = &new_key,
+            .out_error = error,
+            .userdata = fac_userdata,
+         };
+         // Call the callback to create the new key
+         if (!fac (&ctx)) {
+            bsonParseError = "Factory function indicated failure";
+         } else {
+            // Append to the field
+            BSON_APPEND_VALUE (out_field, "keyId", &new_key);
+         }
+         bson_value_destroy (&new_key);
+      }));
+}
+
+/// Generate the "encryptedFields" output for auto-datakeys
+static void
+_init_encryptedFields (bson_t *out_fields,
+                       const bson_t *in_fields,
+                       auto_datakey_factory fac,
+                       void *fac_userdata,
+                       bson_error_t *error)
+{
+   // Ref to one encyrptedField
+   bson_t cur_field;
+   bsonVisitEach (
+      *in_fields,
+      // Each field must be a document element
+      if (not(type (doc)),
+          then (error ("Each 'encryptedFields' element must be a document"))),
+      // Append a new element with the same name as the field:
+      storeDocRef (cur_field),
+      append (
+         *out_fields,
+         kv (
+            bson_iter_key (&bsonVisitIter),
+            // Construct the encryptedField document from the input:
+            doc (do(_init_1_encryptedField (
+               bsonBuildContext.doc, &cur_field, fac, fac_userdata, error))))));
+}
+
+bool
+_mongoc_encryptedFields_fill_auto_datakeys (bson_t *const out_fields,
+                                            const bson_t *const in_fields,
+                                            const auto_datakey_factory factory,
+                                            void *const userdata,
+                                            bson_error_t *const error)
+{
+   BSON_ASSERT_PARAM (in_fields);
+   BSON_ASSERT_PARAM (out_fields);
+   BSON_ASSERT_PARAM (factory);
+
+   if (error) {
+      *error = (bson_error_t){0};
+   }
+   bson_init (out_fields);
```

Comment:
```suggestion
```
Strongly recommend requiring an already-initialized `bson_t` to be passed as `out_fields` instead of initializing within the function to avoid repeated initializations as currently being done in `test_auto_datakeys`:

```c
   bson_t out_fields = BSON_INITIALIZER; // Responsibility of caller.
   bool okay = _mongoc_encryptedFields_fill_auto_datakeys (
      &out_fields, &in_fields, _auto_datakeys, NULL, &error);
   // ...
   bson_destroy (&out_fields); // Also responsibility of caller.
```

Reply:
(See related comment).

I use `BSON_INITIALIZER` not as a "value" for `bson_t`, but as a a `NULL`-ish for `bson_t`, in that it is a non-allocating "zero state" for a `bson_t` that is safe to `bson_destroy` and is safe to overwrite without "freeing" it. (although it isn't really equivalent to a "NULL" for `bson_t`, because it is indistinguishable from an empty allocated document object). Alas, the plight of the sometimes pointer-like `bson_t`. This "design" all bursts into flames with `BSON_MEMCHECK`, which is already fairly broken in other ways. 😢 Having a new `BSON_NULL` initializer that is a no-op on `destroy` could make it more clear what the intent is.